**Challenge 3**

In [312]:

import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
import pandas as pd

#Importamos la información del xsv. 
values=pd.DataFrame(pd.read_csv("Siderophore_DB.csv"))


In [313]:
values.info() #Existen aprox en 72 en nulls en microorganismo.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 221 entries, 0 to 220
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ID                    221 non-null    object
 1   Compounds             221 non-null    object
 2   Microorganism         149 non-null    object
 3   hydroxamate           221 non-null    int64 
 4   catecholate           221 non-null    int64 
 5   a-hydroxycarboxylate  221 non-null    int64 
 6   carboxylate           221 non-null    int64 
 7   phenolate             221 non-null    int64 
 8   citrate               221 non-null    int64 
 9   other                 221 non-null    int64 
 10  SMILES                221 non-null    object
 11  validate smiles       221 non-null    bool  
 12  SMARTS                221 non-null    object
 13  SELFIES               221 non-null    object
dtypes: bool(1), int64(7), object(6)
memory usage: 22.8+ KB


In [314]:
#Imputamos quitando los valores donde existan nulls. 
values= values[values['Microorganism'].notnull()]
values.info()

<class 'pandas.core.frame.DataFrame'>
Index: 149 entries, 0 to 204
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ID                    149 non-null    object
 1   Compounds             149 non-null    object
 2   Microorganism         149 non-null    object
 3   hydroxamate           149 non-null    int64 
 4   catecholate           149 non-null    int64 
 5   a-hydroxycarboxylate  149 non-null    int64 
 6   carboxylate           149 non-null    int64 
 7   phenolate             149 non-null    int64 
 8   citrate               149 non-null    int64 
 9   other                 149 non-null    int64 
 10  SMILES                149 non-null    object
 11  validate smiles       149 non-null    bool  
 12  SMARTS                149 non-null    object
 13  SELFIES               149 non-null    object
dtypes: bool(1), int64(7), object(6)
memory usage: 16.4+ KB


In [315]:
#Creamos una lisat preliminar de features. 

num_grupos = values.drop(["ID","Compounds","Microorganism","SMILES","validate smiles","SMARTS","SELFIES"], axis=1)
num_grupos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 149 entries, 0 to 204
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   hydroxamate           149 non-null    int64
 1   catecholate           149 non-null    int64
 2   a-hydroxycarboxylate  149 non-null    int64
 3   carboxylate           149 non-null    int64
 4   phenolate             149 non-null    int64
 5   citrate               149 non-null    int64
 6   other                 149 non-null    int64
dtypes: int64(7)
memory usage: 9.3 KB


In [316]:
#Creamos una columna que acumule la suma de todos los grupos funcionales. 

size_matrix=num_grupos.shape
valores=[]
for row in range(size_matrix[0]):
    valores.append(num_grupos.iloc[row].sum()) 
num_grupos["Numero grupos funcionales"]=valores

In [364]:
#Separamos dependiendo del numero de grupos funcionales. 
filas_suma_uno = num_grupos[num_grupos["Numero grupos funcionales"] == 1]
filas_suma_dos = num_grupos[num_grupos["Numero grupos funcionales"] == 2]
filas_suma_tres = num_grupos[num_grupos["Numero grupos funcionales"] == 3]
filas_suma_cuatro = num_grupos[num_grupos["Numero grupos funcionales"] == 4]


In [365]:
#Usando las filas donde solo suma un grupo funcional
indices_fila_uno=filas_suma_uno.index
indices_fila_uno


Index([11, 27, 61, 68, 70, 148, 166, 169], dtype='int64')

In [366]:
#El target son los compounds. 
target=values["Compounds"]
target


0          2,3-dihydroxybenzoic acid
1        2,3-dihydroxybenzoylglycine
2         2,3-dihydroxybenzoylserine
3      2,3-dihydroxybenzoylthreonine
4              2-hydroxybenzoic acid
                   ...              
200                    vanchrobactin
201                     vibriobactin
202                     vibrioferrin
203                       vicibactin
204                   yersiniabactin
Name: Compounds, Length: 149, dtype: object

In [367]:
#Codificamos los target.
target_encoder=LabelEncoder().fit_transform(target)
target_encoder

array([  0,   1,   2,   3,   6,   4,   5,   7,   8,  15,  16,  17,  18,
        19,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,  31,
        32,  33,  34,  35,   9,  36,  37,  38,  39,  40,  41,  42,  43,
        44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,
        57,  58,  59,  60,  61,  62,  63,  64,  65,  10,  66,  67,  11,
        68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,
        81,  12,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,
        93,  13,  94,  95,  96,  97, 105,  98,  99, 100, 101, 102, 103,
       104, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121,  14, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148])

In [369]:

#Usando los otros datos. 
values_1=pd.DataFrame(pd.read_csv("Siderophore_DB1.csv"))
#Seleccionando el indice de inicio de las columnas.
index_inicio=values_1.columns.get_loc("fr_Al_COO")
#Seleccionando las columnas feature.  
features_1=values_1.iloc[:,index_inicio:]
features_1


fr_Al_COO  fr_Al_OH  fr_Ar_OH  fr_COO  fr_COO2  fr_C_O  fr_C_O_noCOO   
0            0         0         2       1        1       1             0  \
1            1         0         2       1        1       2             1   
2            1         1         2       1        1       2             1   
3            1         1         2       1        1       2             1   
4            0         0         1       1        1       1             0   
..         ...       ...       ...     ...      ...     ...           ...   
216          0         0         0       0        0       0             0   
217          2         0         2       2        2       5             3   
218          2         0         2       2        2       5             3   
219          2         0         1       2        2       5             3   
220          2         0         2       2        2       5             3   

     fr_NH0  fr_NH1  fr_NH2  fr_N_O  fr_amide  fr_benzene  fr_ester   
0         0       0       0       0         0           1         0  \
1         0       1       0       0         1           1         0   
2         0       1       0       0         1           1         0   
3         0       1       0       0         1           1         0   
4         0       0       0       0         0           1         0   
..      ...     ...     ...     ...       ...         ...       ...   
216       2       1       0       0         0           1         0   
217       4       2       1       0         3           1         0   
218       4       2       1       0         3           1         0   
219       4       2       1       0         3           1         0   
220       4       2       1       0         3           1         0   

     fr_para_hydroxylation  fr_phenol  fr_phenol_noOrthoHbond  
0                        1          2                       1  
1                        1          2                       2  
2                        1          2                       2  
3                        1          2                       2  
4                        1          1                       0  
..                     ...        ...                     ...  
216                      1          0                       0  
217                      0          2                       2  
218                      0          2                       2  
219                      0          1                       1  
220                      0          2                       2  

[221 rows x 17 columns]

Se realizo un primer ensayo con  los cuatro grupos pero en si los valores no eran muchos y además cuando se aplica la clasificación en este caso no es tan efectiva ya que calisifcar casi 148 posibles valores distintos no seria muy acertado. 

In [370]:
#Para la segunda base de datos realizamos lo mismo. 

size_matrix=features_1.shape
nuevos_valores=[]
for row in range(size_matrix[0]):
    nuevos_valores.append(features_1.iloc[row].sum()) 
features_1["Numero grupos funcionales"]=nuevos_valores

In [371]:
features_1

fr_Al_COO  fr_Al_OH  fr_Ar_OH  fr_COO  fr_COO2  fr_C_O  fr_C_O_noCOO   
0            0         0         2       1        1       1             0  \
1            1         0         2       1        1       2             1   
2            1         1         2       1        1       2             1   
3            1         1         2       1        1       2             1   
4            0         0         1       1        1       1             0   
..         ...       ...       ...     ...      ...     ...           ...   
216          0         0         0       0        0       0             0   
217          2         0         2       2        2       5             3   
218          2         0         2       2        2       5             3   
219          2         0         1       2        2       5             3   
220          2         0         2       2        2       5             3   

     fr_NH0  fr_NH1  fr_NH2  fr_N_O  fr_amide  fr_benzene  fr_ester   
0         0       0       0       0         0           1         0  \
1         0       1       0       0         1           1         0   
2         0       1       0       0         1           1         0   
3         0       1       0       0         1           1         0   
4         0       0       0       0         0           1         0   
..      ...     ...     ...     ...       ...         ...       ...   
216       2       1       0       0         0           1         0   
217       4       2       1       0         3           1         0   
218       4       2       1       0         3           1         0   
219       4       2       1       0         3           1         0   
220       4       2       1       0         3           1         0   

     fr_para_hydroxylation  fr_phenol  fr_phenol_noOrthoHbond   
0                        1          2                       1  \
1                        1          2                       2   
2                        1          2                       2   
3                        1          2                       2   
4                        1          1                       0   
..                     ...        ...                     ...   
216                      1          0                       0   
217                      0          2                       2   
218                      0          2                       2   
219                      0          1                       1   
220                      0          2                       2   

     Numero grupos funcionales  
0                           10  
1                           16  
2                           17  
3                           17  
4                            7  
..                         ...  
216                          5  
217                         31  
218                         31  
219                         28  
220                         31  

[221 rows x 18 columns]

In [372]:
#Separamos dependiendo del numero de grupos funcionales. 
filas_suma_uno_1 = features_1[features_1["Numero grupos funcionales"] == 1]
filas_suma_dos_1 = features_1[features_1["Numero grupos funcionales"] == 2]
filas_suma_tres_1 = features_1[features_1["Numero grupos funcionales"] == 3]
filas_suma_cuatro_1 = features_1[features_1["Numero grupos funcionales"] == 4]

In [373]:
#Usando las filas donde solo suma un grupo funcional
filas_suma_dos_1

fr_Al_COO  fr_Al_OH  fr_Ar_OH  fr_COO  fr_COO2  fr_C_O  fr_C_O_noCOO   
60           0         0         0       0        0       0             0  \
61           0         0         0       0        0       0             0   
148          0         0         0       0        0       0             0   
170          0         0         1       0        0       0             0   
213          0         0         1       0        0       0             0   

     fr_NH0  fr_NH1  fr_NH2  fr_N_O  fr_amide  fr_benzene  fr_ester   
60        2       0       0       0         0           0         0  \
61        2       0       0       0         0           0         0   
148       2       0       0       0         0           0         0   
170       1       0       0       0         0           0         0   
213       1       0       0       0         0           0         0   

     fr_para_hydroxylation  fr_phenol  fr_phenol_noOrthoHbond   
60                       0          0                       0  \
61                       0          0                       0   
148                      0          0                       0   
170                      0          0                       0   
213                      0          0                       0   

     Numero grupos funcionales  
60                           2  
61                           2  
148                          2  
170                          2  
213                          2

In [354]:
print(len(target_encoder),len(features_1["Numero grupos funcionales"]))
target_encoder

149 221


array([  0,   1,   2,   3,   6,   4,   5,   7,   8,  15,  16,  17,  18,
        19,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,  31,
        32,  33,  34,  35,   9,  36,  37,  38,  39,  40,  41,  42,  43,
        44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,
        57,  58,  59,  60,  61,  62,  63,  64,  65,  10,  66,  67,  11,
        68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,
        81,  12,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,
        93,  13,  94,  95,  96,  97, 105,  98,  99, 100, 101, 102, 103,
       104, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121,  14, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148])

In [375]:
#Creamos el feature final. 
features_2=features_1[0:149]

In [387]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Dividir los datos en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(features_2, target_encoder, test_size=0.2, random_state=42)

# Crear modelo SVC
modelo = SVC()

# Entrenar el modelo con los datos de entrenamiento
modelo.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred = modelo.predict(X_test)

# Calcular la precisión
precision = accuracy_score(y_test, y_pred)
print("Precisión del modelo SVC:", precision)


Precisión del modelo SVC: 0.0


El modelo no tiene precisión por existir demasiados posibles valores de claisifación. Se utilizo un test size de 20% pero no es tan óptimo. Lo ideal sería aplicar estos casos de clasificación a varaibles que tengan menos posibles valores de target. 